src/Data/Text/Encoding.hs

```diff
@@ -301,6 +296,131 @@ getCodePointStateOrError (Utf8DecodeState mCpSt (bs1@(B.length -> len1) : bss')
     Just (_, cpLen) -> Right cpLen
 getCodePointStateOrError _ = Right 0
 
+#ifdef SIMDUTF
+foreign import ccall unsafe "_hs_text_is_valid_utf8" c_is_valid_utf8
+    :: Ptr Word8 -> CSize -> IO CInt
+#endif
+
+{-
+parseUtf8ChunkFrom chunk from = (n, ms)
+
+n is the end index of the longest prefix of chunk that is valid UTF-8 starting
+from index 'from' where the length of the prefix = n - from
+
+ms
+* When ms = Nothing, there is an error: the bytes from index n and beyond are
+  not part of a valid UTF-8 code point.
+* When ms = Just s, all of the remaining bytes from index n and beyond are the
+  beginning of an incomplete UTF-8 code point, and s is the corresponding
+  intermediate decoding state, which can be used to parse the next chunk with
+  `parseUtf8Chunk`
+-}
+parseUtf8ChunkFrom :: ByteString -> Int -> (Int, Maybe Utf8CodePointState)
+parseUtf8ChunkFrom bs@(B.length -> len) pos =
+  let w n test = (len - pos) >= n && test (B.index bs $ len - n)
+      wi n word8 = w n (>= word8)
+      wc n mask word8 = w n $ (word8 ==) . (mask .&.)
+      guessUtf8Boundary
+        | wi 3 0xf0 = Just $ len - 3  -- third to last char starts a four-byte code point
+        | wi 2 0xe0 = Just $ len - 2  -- pre-last char starts a three-or-four-byte code point
+        | wi 1 0xc2 = Just $ len - 1  -- last char starts a two-(or more-)byte code point
+        | wc 4 0xf8 0xf0 ||           -- last four bytes are a four-byte code point
+          wc 3 0xf0 0xe0 ||           -- last three bytes are a three-byte code point
+          wc 2 0xe0 0xc0 ||           -- last two bytes are a two-byte code point
+          w 1 (< 0x80) = Just len     -- last char is ASCII
+        | otherwise = Nothing         -- no clue
```

Comment:
It's only necessary to check for incomplete code points at the end, and not complete ones. After you don't find an incomplete code point, then the whole bytestring is valid or there is an error, and you can guess `len` as the boundary, which will lead to  the necessary validation anyway.

Reply:
Done.